radicle/src/cob/identity.rs

```diff
@@ -0,0 +1,701 @@
+    /// that hasn't happened yet.
+    #[error("causal dependency {0:?} missing")]
+    Missing(OpId),
+    #[error("the proposal is published")]
+    Published,
+    #[error(transparent)]
+    Publish(#[from] PublishError),
+    #[error("the revision {0:?} is redacted")]
+    Redacted(OpId),
+    /// Error applying an op to the patch thread.
+    #[error("thread apply failed: {0}")]
+    Thread(#[from] thread::OpError),
+}
+
+/// Error publishing the proposal.
+#[derive(Error, Debug)]
+pub enum PublishError {
+    #[error(transparent)]
+    Identity(#[from] IdentityError),
+    #[error("the proposal {0} is closed")]
+    Closed(OpId),
+    #[error("the revision {0} is missing")]
+    Missing(OpId),
+    #[error("the revision {0} is already published")]
+    Published(OpId),
+    #[error("the revision {0} is redacted")]
+    Redacted(OpId),
+    #[error(transparent)]
+    Doc(#[from] DocError),
+    #[error("signatures did not reach quorum threshold: {0}")]
+    Quorum(usize),
+}
+
+/// Error updating or creating proposals.
+#[derive(Error, Debug)]
+pub enum Error {
+    #[error("apply failed: {0}")]
+    Apply(#[from] ApplyError),
+    #[error("store: {0}")]
+    Store(#[from] store::Error),
+}
+
+/// Propose a new [`Doc`] for an [`Identity`]. The proposal can be
+/// reviewed by gathering [`Signature`]s for accepting the changes, or
+/// rejecting them.
+///
+/// Once a proposal has reached the quourum threshold for the previous
+/// [`Identity`] then it may be published to the person's local
+/// storage using [`Proposal::publish`].
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct Proposal {
+    /// Title of the proposal.
+    title: LWWReg<Max<String>>,
+    /// Description of the proposal.
+    description: LWWReg<Max<String>>,
+    /// Current state of the proposal.
+    pub state: LWWReg<Max<State>>,
+    /// List of revisions for this proposal.
+    revisions: GMap<RevisionId, Redactable<Revision>>,
+}
+
+#[derive(Debug, Clone, Copy, Default, PartialEq, Eq, Serialize, Deserialize)]
+#[serde(rename_all = "camelCase")]
+pub enum State {
+    #[default]
+    Open,
+    Closed,
+    Published,
+}
+
+impl PartialOrd for State {
+    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
+        Some(self.cmp(other))
+    }
+}
+
+impl Ord for State {
+    fn cmp(&self, other: &Self) -> Ordering {
+        match (self, other) {
+            (State::Open, State::Open) => Ordering::Equal,
+            (State::Closed, State::Closed) => Ordering::Equal,
+            (State::Published, State::Published) => Ordering::Equal,
+            (State::Open, State::Closed) => Ordering::Less,
+            (State::Open, State::Published) => Ordering::Less,
+            (State::Closed, State::Open) => Ordering::Greater,
+            (State::Closed, State::Published) => Ordering::Less,
+            (State::Published, State::Open) => Ordering::Greater,
+            (State::Published, State::Closed) => Ordering::Greater,
+        }
```

Comment:
If so, I'd probably also include a test to prevent the order being changed by mistake.

Reply:
Ya, I was ensuring the order so that it doesn't get changed by reordering the fields. Why rely on a test for this, out of interest?

Reply:
My thought is that the above is better expressed by ordering the fields in the enum, as it's much more readable and less prone to error.

However, since other developers may not realize that the order matters, we should also have a test that just has like `asser!(x < y && y < z)` etc.

Reply:
We do exactly this here for example, since the order of the fields matter: https://github.com/radicle-dev/heartwood/blob/master/radicle/src/cob/issue.rs#L478

Reply:
Cool, thanks for the explanation. I was thinking more along the lines of "The Code is Law", but happy to change it to this instead :)

Reply:
Yeah, it's just one of those few cases where field order has a real consequence, and I can imagine someone organizing fields alphabetically breaking everything without realizing, if we don't have a test :)